<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [362]:
import pandas as pd
import psycopg2

In [363]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [364]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [365]:
# текст запроса
query_3_1 = f'''select count(*)
                from public.vacancies
'''

In [366]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

/tmp/ipykernel_4429/1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [367]:
# текст запроса
query_3_1 = f'''select count(*)
                from public.employers
'''

In [368]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

/tmp/ipykernel_4429/1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [369]:
# текст запроса
query_3_1 = f'''select count(*)
                from public.areas
'''

In [370]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

/tmp/ipykernel_4429/1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [371]:
# текст запроса
query_3_1 = f'''select count(*)
                from public.industries
'''

In [372]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

/tmp/ipykernel_4429/1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0    294

***

# выводы по предварительному анализу данных
По результатам запросов видим, что в базе данных содержится более 49 тыс. вакансий, работодателями по которым может являться 23501 компания.
БД охватывает информацию из 1362 регионов и в 294 сферах деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [373]:
# текст запроса
cnt = f'''select 
        count(vac.id),
        ar.name
        from public.vacancies as vac
        join public.areas as ar on vac.area_id=ar.id
        group by ar.id
        order by count(vac.id) desc
        limit 15
'''

In [374]:
# результат запроса
df = pd.read_sql_query(cnt, connection)
df

/tmp/ipykernel_4429/806200914.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(cnt, connection)


count             name
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
5    1698     Екатеринбург
6    1670  Нижний Новгород
7    1415           Казань
8    1301        Краснодар
9    1144           Самара
10   1131   Ростов-на-Дону
11   1063          Воронеж
12    851          Ташкент
13    847       Красноярск
14    840       Нур-Султан

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [375]:
# текст запроса
salary = '''select
        count(*)
        from public.vacancies
        where salary_from is not null or salary_to is not null
'''

In [376]:
# результат запроса
df = pd.read_sql_query(salary, connection)
display(df)

/tmp/ipykernel_4429/689179189.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(salary, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [377]:
# текст запроса
sal_avg_min = '''select
                avg(salary_from) as avg_min,
                avg(salary_to) as avg_max                
                from public.vacancies
'''

In [378]:
# результат запроса
df = pd.read_sql_query(sal_avg_min,connection)
display(round(df,0))

/tmp/ipykernel_4429/3354079060.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sal_avg_min,connection)


avg_min   avg_max
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [379]:
# текст запроса
vac_count = '''select
        schedule,
        employment,
        count(*)
        from public.vacancies
        group by(schedule, employment)
        order by count(*) desc
'''

In [380]:
# результат запроса
df = pd.read_sql_query(vac_count,connection)
display(df)
display(df.iloc[1])

/tmp/ipykernel_4429/4023093829.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(vac_count,connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

schedule      Удаленная работа
employment    Полная занятость
count                     7802
Name: 1, dtype: object

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [381]:
# текст запроса
exp = '''select
        experience
        from public.vacancies
        group by(experience)
        order by count(experience)
'''

In [382]:
# результат запроса
df = pd.read_sql_query(exp, connection)
df

/tmp/ipykernel_4429/2839512516.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(exp, connection)


experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

# выводы по детальному анализу вакансий
По результатам детального анализа вакансий видим, что первая пятерка регионов по вакансиям - это Москва, СПб, Минск, Новосибирск, Алматы.
Причем вакансий в Москве почти в 2 раза больше, чем в Санкт-Петербурге (1 и 2 места соответственно). По мере убывания разница становится менее существенной.
Хотя бы одно из двух полей с зарплатой заполнено в менее, чем половине вакансий - в 24073. И средние значения для нижней и верхней границы зарплатной вилки составляют 71065 руб. и 110537 руб. соответственно.
Можно предположить, что существенный вклад в размер средней зарплаты вносит Москва. В дополнительном исследовании можно проверить это утверждение, посчитав
средние значения для нижней и верхней границы зарплатной вилки для Москвы и других регионов.

Самыми востребованными с существенным преобладанием являются работники на полный день и полную занятость. Далее идет удаленная работа и гибкий график, оба с полной занятостью.

В БД размещено больше всего вакансий с требуемым опытом работы от 1 до 3 лет. Меньше всего вакансий для опыта более 6 лет. 
Популярность опыта работы от 1 до 3 лет объясняется скорее всего тем, что требуются работники на младшие должности.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [383]:
# текст запроса
empl = '''select
        emp.name,
        count(vac.id)
        from public.vacancies vac
            join public.employers emp on emp.id=vac.employer_id
        group by (emp.id)
        order by count(vac.id) desc
'''

In [384]:
# результат запроса
df = pd.read_sql_query(empl, connection)
display(df.iloc[[0,4]])

/tmp/ipykernel_4429/3253602763.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(empl, connection)


name  count
0         Яндекс   1933
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [385]:
# текст запроса
area_count_empl = '''select
                    ar.name area_name,
                    count(distinct emp.id) cnt_employers,
                    count(vac.id) cnt_vacations
                    from public.employers emp
                    left join public.vacancies vac on emp.area=vac.area_id
                    join public.areas ar on ar.id=emp.area
                    where vac.id is null
                    group by ar.id
                    order by count(distinct emp.id) desc
                    limit 1                    
'''


In [386]:
# результат запроса
df = pd.read_sql_query(area_count_empl, connection)
df

/tmp/ipykernel_4429/2572618035.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(area_count_empl, connection)


area_name  cnt_employers  cnt_vacations
0    Россия            410              0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [387]:
# текст запроса
cnt_areas = '''select
                emp.name employer_name,
                count(distinct vac.area_id) cnt_areas
                from public.vacancies vac
                join public.employers emp on vac.employer_id=emp.id
                group by(emp.id)
                order by count(distinct vac.area_id) desc
'''

In [388]:
# результат запроса
df = pd.read_sql_query(cnt_areas, connection)
df

/tmp/ipykernel_4429/1414662658.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(cnt_areas, connection)


employer_name  cnt_areas
0                                      Яндекс        181
1                                  Ростелеком        152
2                                  Спецремонт        116
3                      Поляков Денис Иванович         88
4                                   ООО ЕФИН          71
...                                       ...        ...
14901                            НПП Авиатрон          1
14902              Центр дистанционных торгов          1
14903  Городские Телекоммуникационные Системы          1
14904                       Введенский, Отель          1
14905              Группа Компаний Автоимпорт          1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [389]:
# текст запроса
cnt_emp_ind = '''select
                count(*)
                from public.employers emp
                left join public.employers_industries ind on emp.id=ind.employer_id
                where ind.industry_id is null
'''

In [390]:
# результат запроса
df = pd.read_sql_query(cnt_emp_ind, connection)
df

/tmp/ipykernel_4429/3591177189.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(cnt_emp_ind, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [391]:
# текст запроса
emp_four_ind = '''select
                emp.name
                from public.employers emp
                join public.employers_industries ind on emp.id=ind.employer_id
                group by(emp.id)
                having count(emp.id)=4
                order by emp.name
                offset 2 limit 1
'''

In [392]:
# результат запроса
df = pd.read_sql_query(emp_four_ind, connection)
df

/tmp/ipykernel_4429/2952286381.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(emp_four_ind, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [393]:
# текст запроса
cnt_it_ind = '''select 
                count(*)
                from public.industries ind
                join public.employers_industries emp on emp.industry_id=ind.id
                where ind.name = 'Разработка программного обеспечения'              
'''

In [394]:
# результат запроса
df = pd.read_sql_query(cnt_it_ind, connection)
df

/tmp/ipykernel_4429/127827568.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(cnt_it_ind, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [395]:
# код для получения списка городов-милионников
import requests 
from bs4 import BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
links = page.find_all('a')
list_c = [link.text for link in links[52:68]]
print(list_c)


['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [396]:
# текст запроса
ya_emp = '''select
            ar.name,
            count(vac.id)
            from public.vacancies vac
            join public.areas ar on vac.area_id=ar.id
            join public.employers emp on vac.employer_id=emp.id
            where emp.name='Яндекс' and ar.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
            group by ar.id
            
            union
            
            select 'Total',
            count(vac.id)
            from public.vacancies vac
            join public.areas ar on vac.area_id=ar.id
            join public.employers emp on vac.employer_id=emp.id
            where emp.name='Яндекс' and ar.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
          
            order by 2
'''

In [397]:
# результат запроса
df = pd.read_sql_query(ya_emp, connection)
df

/tmp/ipykernel_4429/1235538186.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(ya_emp, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

# выводы по анализу работодателей
По результатам запросов видим, что на 1 месте по количеству размещенных вакансий находится Яндекс. У компании размещено 1933 вакансии. При этом у Газпром нефти размещена 331 вакансия. Она находится на 5 месте.

Яндекс также является работодателем, который размещает вакансии в наибольшем количестве регионов. По результатам запроса таких регионов 181. На 2 месте Ростелеком, который размещает вакансии в 152 регионах.

Всего работодателей, которые публикуют вакансии хотя бы в 1 регионе, 14906. При этом в БД присутствуют регионы, в которых есть работодатели, но нет вакансий.

Не все работодатели указывают сферу деятельности. В базе 8419 компаний, указавших эту информацию. Некоторые работодатели определяют несколько сфер деятельности, например, фирма 2ГИС указала 4. Разработка программного обеспечения указана как сфера деятельности у 3553 работодателей.

У компании Яндекс размещено 485 вакансий в городах-миллионниках, больше всего в Мск - 54, но и в остальных городах спрос на сотрудников высок.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [398]:
# текст запроса
cnt_dat='''select
            count(*)
            from public.vacancies
            where name ilike '%data%' or name ilike '%данн%'
'''

In [399]:
# результат запроса
df = pd.read_sql_query(cnt_dat, connection)
df

/tmp/ipykernel_4429/1447535689.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(cnt_dat, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [400]:
# текст запроса
ds_junior_vac = '''select
                   count(*)
                   from public.vacancies
                   where (name ilike any ('
	                {%data scientist%, %data science%, %исследователь данных%, 
	                %machine learning%, %machine learning%, %машинн%обучен%}
	                ') or 
	                (name not ilike '%HTML%' and name like '%ML%')) and
                    (name ilike '%junior%' or 
                    experience ilike '%Нет опыта%' or
                    employment ilike '%Стажировка%')
'''

In [401]:
# результат запроса
df = pd.read_sql_query(ds_junior_vac, connection)
df

/tmp/ipykernel_4429/3418477352.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(ds_junior_vac, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [402]:
# текст запроса
ds_vac_sql = '''select
                   count(*)
                   from public.vacancies
                   where (name ilike any ('
	                {%data scientist%, %data science%, %исследователь данных%, 
	                %machine learning%, %machine learning%, %машинн%обучен%}
	                ') or 
	                (name not ilike '%HTML%' and name like '%ML%')) and
                    (key_skills ilike any ('{%SQL%, %postgres%}'))
'''

In [403]:
# результат запроса
df = pd.read_sql_query(ds_vac_sql, connection)
df

/tmp/ipykernel_4429/3060917210.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(ds_vac_sql, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [404]:
# текст запроса
ds_vac_pth = '''select
                   count(*)
                   from public.vacancies
                   where (name ilike any ('
	                {%data scientist%, %data science%, %исследователь данных%, 
	                %machine learning%, %machine learning%, %машинн%обучен%}
	                ') or 
	                (name not ilike '%HTML%' and name like '%ML%')) and
                    (key_skills ilike '%Python%')
'''

In [405]:
# результат запроса
df = pd.read_sql_query(ds_vac_pth, connection)
df

/tmp/ipykernel_4429/2997831170.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(ds_vac_pth, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [406]:
# текст запроса
cnt_key_skills = '''select
                    avg((length(key_skills) - length(replace(key_skills, '\t', ''))+1))
                    from public.vacancies
                    where (name ilike any ('
	                {%data scientist%, %data science%, %исследователь данных%, 
	                %machine learning%, %machine learning%, %машинн%обучен%}
	                ') or 
	                (name not ilike '%HTML%' and name like '%ML%')) 
                    '''

In [407]:
# результат запроса
df = pd.read_sql_query(cnt_key_skills, connection)
round(df,2)

/tmp/ipykernel_4429/4080239199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(cnt_key_skills, connection)


avg
0  6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [408]:
# текст запроса
avg_salary = '''select
                experience,
                round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)),0)
                from public.vacancies
                where (name ilike any ('
	            {%data scientist%, %data science%, %исследователь данных%, 
	            %machine learning%, %machine learning%, %машинн%обучен%}
	            ') or 
	            (name not ilike '%HTML%' and name like '%ML%'))
                group by experience
                '''

In [409]:
# результат запроса
df = pd.read_sql_query(avg_salary, connection)
df

/tmp/ipykernel_4429/4122466633.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(avg_salary, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  139675.0
3       От 3 до 6 лет  243115.0

***

# выводы по предметному анализу
По результатам запросов в базе есть 1771 вакансия, в названии которых содержатся слова 'data' или 'данн'. При этом для начинающего дата-сайентиста есть 51 вакансия. Вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres, содержится в базе 201. А Python указан как ключевой навык у DS в 351 случае.
В среднем в вакансиях для DS указывают 6,41 ключевых навыков.
Самую высокую среднюю зарплату для DS указывают для опыта работы от 3 до 6 лет. Это сумма составляет 243115 рублей. Для соискателей с опытом работы от 1 до 3 лет зарплата в среднем составит 139675 рублей.

# Общий вывод по проекту

# подведем итог исследования, обобщите выводы

По результатам анализа видим, что в базе данных представлены 49197 вакансий, 23501 работодатель и охвачена информация из 1362 регионов и в 294 сферах деятельности.

По количеству размещенных вакансий лидируют Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. В Москве (1 место) вакансий почти в 2 раза больше, чем в СПб (2 место), далее разница сглаживается.

Менее половины вакансий содержит информацию о заработной плате. И средние значения для нижней и верхней границы зарплатной вилки составляют 71 тыс. рубей и 110 тыс. рублей соответственно. Можно проанализировать среднюю предлагаемую зарплату в Москве и других регионах. Запрос представлен ниже. Видим, что зарплата в Москве ~ на 50 тыс. выше, чем средняя во всех других регионах, но существенно на общую среднюю это не оказывает существенного влияния.

Самыми востребованными являются работники на полный день и полную занятость. Преобладает опыт работы от 1 до 3 лет.

Работодателем, разместившим больше всего вакансий, является Яндекс. Он также охватывает наибольшее количество регионов - 181. Также у Яндекс в городах-миллионниках размещено 485 вакансий, больше всего в Мск - 54, но и в остальных городах спрос на сотрудников высок.

Для начинающего дата-сайентиста в БД есть 51 вакансия.

SQL или postgres востребован как ключевой навык в вакансиях для DS в 201 случае, а Python - в 351.

Самую высокую среднюю зарплату для DS указывают для опыта работы от 3 до 6 лет. Это сумма составляет 243115 рублей, для соискателей с опытом работы от 1 до 3 лет зарплата в среднем составляет 139675 рублей.


In [424]:
sal_avg = '''select
            'Москва' as name_area,
            round(avg(salary_from),0) as avg_min,
            round(avg(salary_to),0) as avg_max                
            from public.vacancies vac
            join public.areas ar on vac.area_id=ar.id
            where ar.name = 'Москва'
            
            union

            select
            'other',
            round(avg(salary_from),0) as avg_min,
            round(avg(salary_to),0) as avg_max                
            from public.vacancies vac
            join public.areas ar on vac.area_id=ar.id
            where ar.name <> 'Москва'                     
'''


In [425]:
# результат запроса
df = pd.read_sql_query(sal_avg, connection)
df

/tmp/ipykernel_4429/2346992945.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sal_avg, connection)


name_area   avg_min   avg_max
0     other   68270.0  106375.0
1    Москва  113409.0  160594.0